File: leetcode/python/LEET1038.py
from typing import *
import collections

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

val = 0
def bstToGst(root: TreeNode) -> TreeNode:
    global val
    if root is None:
        return 
    
    bstToGst(root.right)
    val += root.val
    root.val = val
    bstToGst(root.left)

    return root

def solve():
    root = build_tree([4,1,6,0,2,5,7,None,None,None,3,None,None,None,8])
    bstToGst(root)
    print_tree(root)

def build_tree(array: List[int]) -> TreeNode:
    if not array:
        return None

    root = TreeNode(array[0])
    q = collections.deque([root])
    index = 1

    while index < len(array):
        node = q.popleft()
        if array[index] is not None:
            left = TreeNode(array[index])
            node.left = left
            q.append(left)
        index += 1
        
        if index >= len(array): break

        if array[index] is not None:
            right = TreeNode(array[index])
            node.right = right
            q.append(right)
        index += 1

    return root

def print_tree(root):
    q = collections.deque([root])

    while q:
        node = q.popleft()
        if node:
            print(node.val, end=" ")
        else:
            print(None, end=" ")
            continue
        
        q.append(node.left)
        q.append(node.right)
    print()
solve()

File: 0_3.leetcode/java/LEET1038.java
import java.util.*;

public class LEET1038 {
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    public static void main(String[] args) {
        TreeNode root = buildTree(new Integer[] {4,1,6,0,2,5,7,null,null,null,3,null,null,null,8});
        printTree(root);
        bstToGst(root);
        printTree(root);
    }
    static int val = 0;
    public static TreeNode bstToGst(TreeNode root) {
        if(root != null){
            bstToGst(root.right);
            val += root.val;
            root.val = val;
            bstToGst(root.left);
        }
        return root;
    }
    public static void printTree(TreeNode root){
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        while(!q.isEmpty()){
            TreeNode node = q.poll();
            if(node == null){
                System.out.print("NULL ");
                continue;
            }
            System.out.print(node.val +" ");
            
            q.add(node.left);
            q.add(node.right);
        }
        System.out.println();
    }
    public static TreeNode buildTree(Integer[] nums){
        if(nums.length == 0){
            return null;
        }
        Queue<TreeNode> q = new LinkedList<>();
        TreeNode root = new TreeNode(nums[0]);
        int index = 1;
        q.add(root);
        while(index < nums.length){
            TreeNode node = q.poll();
            if(nums[index] != null){
                node.left = new TreeNode(nums[index]);
                q.add(node.left);
            }
            index++;

            if(index >= nums.length)
                break;
            
            if(nums[index] != null){
                node.right = new TreeNode(nums[index]);
                q.add(node.right);
            }
            index++;
        }
        return root;
    }
}
